Change bootstrapper to install the compiled version of KoreBuild

natemcmaster opened the pull request:
Changes the bootstrappers to use the "compiled" KoreBuild. Part of https://github.com/aspnet/KoreBuild/issues/258

This will also override the build.cmd/build.sh in individual repos to use the build.ps1 in the Unverse repo. This is temporary until we've updated all individual repos to use the new bootstrapper.

pakrym commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+        $remotePath = "$ToolsSource/korebuild/artifacts/$version/korebuild.$version.zip"
+
+        try {
+            if ($PSVersionTable.PSVersion -ge '5.0.0.0') {
```
Get-Command -Name "Expand-Archive" -ErrorAction SilentlyContinue

pakrym commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+#>
+[CmdletBinding(PositionalBinding = $false)]
+param(
+    [Alias('p')]
```
Would this conflict with `/p:Property=Value` msbuild argument?

natemcmaster replied:
No. I've verified this works: `build.ps1 -p ../OtherRepo /p:Prop=value`

pakrym commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+        __fetch "$ToolsSource/korebuild/channels/$Channel/latest.txt" $lockFile
     }
+
+    $version = Get-Content $lockFile -Tail 1
```
Do we need a way to pass it in or make it fixed in the repo?

natemcmaster replied:
We intend to commit this lockfile to source when we branch or need to pin for other reasons. Would a `$KoreBuildVersion` flag also be useful?

pakrym replied:
We can always add it later.

pakrym commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+# Functions
+#
+
+function __get_korebuild {
```
nit: Does not look like a powershell naming scheme

pakrym commented:
Poweshell LGTM, @anurse @halter73 please take a look too if you have time.

natemcmaster commented:
Also, if this wasn't clear, these bootstrapper scripts is what I will end up coping into all individual repos.

pakrym commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
     }
+
+    $version = Get-Content $lockFile -Tail 1
+    $korebuildPath = __join_paths $DotNetHome ('buildtools', 'korebuild', $version)
```
Why isn't channel included here? Is it included in version?

natemcmaster replied:
Channel is only used to fetch a specific version.

Example:  https://aspnetcore.blob.core.windows.net/buildtools/korebuild/channels/dev/latest.txt

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+        curl --retry 10 -sSL -f --create-dirs -o $local_path $remote_path || failed=true
+    elif __machine_has wget; then
+        wget --tries 10 -O $local_path $remote_path || failed=true
+    else
```
Check these at the top of the file (before arg parsing). You can just duplicate the check a little (`if ! __machine_has curl && ! __machine_has wget; then ...`, double check the precedence there though I'm not positive what order those are invoked ;))

natemcmaster replied:
Updated

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
 
-    unzip -q -d $tempFolder $localZipFile
+        tmpfile="$(mktemp)"
+        echo "Downloading KoreBuild $version"
```
Colors maybe? You can use something like https://gist.github.com/elucify/c7ccfee9f13b42f11f81 (there are many others out there if you search for 'bash color variables' :)

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+        echo "Downloading KoreBuild $version"
+        __fetch $remote_path $tmpfile
+        unzip -q -d "$korebuild_path" $tmpfile
+        rm $tmpfile || :
```
What's the `:` do here? I'm not familiar.

anurse replied:
Prevent `pipefail`?

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+        -\?|-h|--help)
+            __usage
+            ;;
+        -c|--channel|-Channel)
```
`-Channel` (and similar) are for equivalent syntax to PowerShell? Comment please :)

natemcmaster replied:
Yes, these are meant to align with powershell. Inspiration: dotnet-install.ps1/sh does the same thing. Saves me now and then when muscle memory does the wrong thing.

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+while [[ $# > 0 ]]; do
+    case $1 in
+        -\?|-h|--help)
+            __usage
```
We should exit with `0` exit code in this case. `__usage` should take a flag or variable for the exit code.

natemcmaster replied:
updated

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+
+get_korebuild
+install_tools "$tools_source" "$DOTNET_HOME"
+invoke_repository_build "$repo_path" $@
```
same comment as in PowerShell, can we reduce the "public API" surface here to just one function?

anurse commented on build/RepositoryBuild.proj:
```diff
@@ -60,20 +60,14 @@
-    <Copy SourceFiles="@(_KorebuildItems)" DestinationFolder="$(BuildRepositoryRoot).build\%(RecursiveDir)" SkipUnchangedFiles="true" UseHardlinksIfPossible="true" />
+    <!-- Copy Korebuild lock file to individual repos to align version if the repo doesn't already have one -->
+    <Message Text="Copying KoreBuild lockfile from Universe to repository $(BuildRepositoryRoot)"/>
+    <Copy SourceFiles="$(RepositoryRoot)korebuild-lock.txt" DestinationFolder="$(BuildRepositoryRoot)" />
```
Does KoreBuild write it's version number to the `artifacts` directory? If not, perhaps we should do that too in order to make sure we know what version built a particular set of artifacts.

natemcmaster replied:
korebuild.$version.zip includes a `.version` file inside it that is identical to the `korebuild-lock.txt` file. 

See https://github.com/aspnet/BuildTools/tree/dev/sdk/KoreBuild#layout

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+        Get-RemoteFile "$ToolsSource/korebuild/channels/$Channel/latest.txt" $lockFile
     }
+
+    $version = Get-Content $lockFile -Tail 1
```
I do think a version option would be a good idea. It seems trivial to add here but if it's more than just adding the switch and a couple lines, then it can wait.

natemcmaster replied:
It's more than a few lines. We'd need to make the expected lockfile an artifact available by version too. The lockfile includes more than just the version. Currently it includes the commithash of aspnet/BuildTools which will be important for pinning.

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+    $version = Get-Content $lockFile -Tail 1
+    $korebuildPath = Join-Paths $DotNetHome ('buildtools', 'korebuild', $version)
+
+    if (!(Test-Path $korebuildPath)) {
```
Can we add a switch to force re-download of the version even if it exists? -ResetKoreBuild?

natemcmaster replied:
Can you clarify why you mean by redownload? There are two redownloads possible: (1) the download that fetches the `korebuild-lock.txt` file from the channel, and (2) the re-download of the korebuild artifact. The `-Update` addresses (1) -- it overwrite the local version of korebuild-lock.txt. 
(2) shouldn't be necessary. Once a version is determined from the lockfile, the appropriate korebuild artifact will always be downloaded if it is not already present in `$DOTNET_HOME/buildtools/korebuild/$version`.

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+    $korebuildPath = Join-Paths $DotNetHome ('buildtools', 'korebuild', $version)
+
+    if (!(Test-Path $korebuildPath)) {
+        Write-Host "Downloading KoreBuild $version"
```
Adding a little color to the KoreBuild messages helps them stand out from the MSBuild messages. It might be a nice addition here.

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+        try {
+            if (Get-Command -Name 'Expand-Archive' -ErrorAction Ignore) {
+                # Use built-in commands where possible as they are cross-plat compatible
+                $tmpfile = "$(New-TemporaryFile).zip"
```
[This calls `Path.GetTempFileName`](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs#L26) which creates the file it returns in advance. If `Expand-Archive` (and `ZipFile` below) don't care about the extension, we could just use it without creating extra files (though granted, they are 0-byte).

Alternatively, the way you create the temp file name in the other branch is totally safe, so there's no reason not to just do it here.

Wish we had `Path.GetTempFileName(string extension)`.

natemcmaster replied:
Regrettable, `Expand-Archive` does care. It has to be .zip :-( I just use the alternative creating of the temp file in both places.

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
 }
 
-&"$buildFile" @args
+function Join-Paths($path, $childPaths) {
```
I hate that this isn't standard... Put type annotations on the params though ;)

anurse commented on build.ps1:
```diff
@@ -1,66 +1,140 @@
+    $korebuildPath = Get-KoreBuild
+    Import-Module -Force -Scope Local (Join-Path $korebuildPath 'KoreBuild.psd1')
+    Install-Tools $ToolsSource $DotNetHome
+    Invoke-RepositoryBuild $Path @MSBuildArgs
```
Can these steps be collapsed? Changing the template is a pain and involves changing every repo, I'd like the build script to be as rock-solid and flexible as possible and just delegate everything to a script inside KoreBuild as soon as possible to shift logic to the place we can centrally-update.

natemcmaster replied:
This was very deliberate. Let me explain my thinking:
 - make it clearer to those reading the script which steps will happen when build.ps1 executes.
 - `Install-Tools` and `Invoke-RepositoryBuild` should be treated like an "API". Any changes to arguments or behavior is 'breaking' and should not be done lightly. These will rarely if ever change, so our bootstrappers can delegate the majority of stuff to these two functions.
 - I'm totally okay with one extra line in build.ps1.

FYI - I'm also thinking forward to what we discussed in our runtime-engineering meeting last week: https://github.com/aspnet/specs/pull/77. One of the problems we have to day is that our build is a mix of console tools, tasks, and scripts, scattered across nuget packages, source code, and stuff that gets downloaded on the fly. We want KoreBuild capture the "state" of build tools. Decomposing KoreBuild into a module (KoreBuild.ps**m**1) gives us more flexibility to write scripts that re-use tools from KoreBuild. `Korebuild.ps1` was too inflexible. For example, I want to add a function to KoreBuild.psd1 to `Push-Packages`. This could delegate to `nuget.exe`, `dotnet nuget`, or `PackagePublisher.dll`, but it doesn't matter to the caller since all they need is the powershell function.

anurse commented on build.sh:
```diff
@@ -1,46 +1,141 @@
+        echo "Downloading KoreBuild $version"
+        __get_remote_file $remote_path $tmpfile
+        unzip -q -d "$korebuild_path" $tmpfile
+        rm $tmpfile || :
```
What's the `:` do here? I'm not familiar with it. Does it prevent pipefail from triggering? Comment please (since we are mostly `sh`noobs)

natemcmaster replied:
`:` is a builtin equivalent to `true`. It makes this line is functionally equivalent to `rm $tmpfile >/dev/null 2>&1`, except that we'll get some console output indicating there was a problem.

Try this:
```sh
rm some_bogus_file || :
echo $?
```
It should return 0, but still show you that some_bogus_file did not exist.


FYI - https://unix.stackexchange.com/questions/31673/what-purpose-does-the-colon-builtin-serve

natemcmaster commented:
Improvements to build infrastructure are being delayed until after we branch for 2.0.0. I've captured this feedback and ported it back to aspnet/BuildTools: https://github.com/aspnet/BuildTools/commit/d5fd2ee2fc23117ed0978761434d23e3d44708c2

Closing for now. We can re-open this later this summer.

natemcmaster commented:
:up: :date: - re opening now that we have branched for 2.1.0. Since last review, I've update the bootstrappers a little.

Another important change: because this version of KoreBuild does not do pre-flight restore, any "PackageReference" in repo.targets will not be restored. This means PackagePublisher is not available from the default KoreBuild lifecycle. Instead, I've added a new file `push.cmd` that will push the packages after build is done.

You can try a dry-run push locally by running:
```
push.cmd -WhatIf
```

anurse commented:
Is there a reason why `build.ps1` was renamed to `run.ps1`?

pranavkm commented:
>  I've added a new file push.cmd that will push the packages after build is done.

Where does it push to? Why do we need something in addition to the package publisher we already have?

natemcmaster commented:
> Another important change: because this version of KoreBuild does not do pre-flight restore, any "PackageReference" in repo.targets will not be restored. This means PackagePublisher is not available from the default KoreBuild lifecycle. Instead, I've added a new file push.cmd that will push the packages after build is done.

pranavkm commented on build/RepositoryBuild.targets:
```diff
@@ -71,13 +65,6 @@
-    <Message Text="Publishing the following packages to the volatile feed: @(RepositoryNupkgs -> '%(Filename)%(Extension)', ', ')"
-      Condition="'$(PublishPackages)'=='true' AND '@(RepositoryNupkgs)' != ''" />
-
-    <Exec
```
What happened to this?

natemcmaster commented on run.ps1:
```diff
@@ -0,0 +1,216 @@
+    if ($CommandName -eq 'build') {
+        Invoke-RepositoryBuild $Path @MSBuildArgs
+    }
+    elseif ($CommandName -eq 'push') {
```
Look here.

pranavkm commented:
Wasn't the plan to make it a task?

natemcmaster commented:
It was until I realized that it would introduce binary compatibility issues due to how MSBuild loads tasks. PackagePublisher currently uses NuGet 4.0.0. When I made PackagePublisher a task, it fails to load because the SDK bundles NuGet 4.3.0 which is binary incompatible. Rather than tangle with MSBuild, we can push packages from console/shell only. This is what the `Push-NuGetPackage` cmdlet does under the covers.

natemcmaster commented:
FYI - an alternative I discussed with @pakrym this morning was that instead of push.cmd, we leave the bootstrapper the same and add `/t:PushPackages` instead. We can bundle PackagePublisher in KoreBuild and add a KoreBuild target that pushes all \*.nupkg from $(BuildDir)

natemcmaster commented:
Scratch all that. Changed my mind. We have enough dependencies on NuGet anyways (like NuGetPackageVerifier), so we're unlikely to get much value from the powershell-only approach. Returning this PR to the original approach. Namely:

 -  PackagePublisher is now a built-in task in KoreBuild (https://github.com/aspnet/BuildTools/pull/306)
 - This PR uses the standard bootstrapper build.sh and build.ps1 scripts as found here: https://github.com/aspnet/BuildTools/tree/dev/scripts
 - Added a push.cmd file, which is just a short cut for `build.ps1 /t:PushPackages`

natemcmaster commented:
:bell: I've squashed and rebased this.

pranavkm commented:
Not sure about others, but I'm a bit busy with 2.0 stuff today so unlikely I'd have a chance to look at this. Would be good to have @anurse \ @pakrym have a look at this before this goes in.

pranavkm commented on build.ps1:
```diff
@@ -1,66 +1,177 @@
             }
         }
+        catch {
+            remove-item -Recurse -Force $korebuildPath -ErrorAction Ignore
```
Casing consistency `Remove-Item` (in the finally block too)

pranavkm commented on build.ps1:
```diff
@@ -1,66 +1,177 @@
+    [string]$ToolsSource,
+    [Alias('u')]
+    [switch]$Update,
+    [string]$ConfigFile = (Join-Path $PSScriptRoot 'version.props'),
```
Can we rename the file to `xml` across repos if we intend to treat it as xml and not an msbuild evaluated file?

natemcmaster replied:
Fine with me.

pranavkm commented on build.ps1:
```diff
@@ -1,66 +1,177 @@
 
-    DownloadWithRetry -url $koreBuildZip -downloadLocation $localZipFile -retries 6
+if (!$Channel) { $Channel = 'dev' }
+if (!$ToolsSource) { $ToolsSource = 'https://aspnetcore.blob.core.windows.net/buildtools' }
```
Why not set these inline when they're defined?

natemcmaster replied:
To preserve this order of precedence:
1. Explicit CLI flag. `./build.ps1 -ToolsSource C:\dev\aspnet\BuildTools\artifacts`
2. Path in version.props `<KoreBuildToolsSource>https://something/else/</KoreBuildToolsSource>`
3. If neither 1 or 2, fallback to https://aspnetcore.blob.core.windows.net/buildtools

pranavkm commented on build.sh:
```diff
@@ -1,46 +1,195 @@
+    elif __machine_has wget; then
+        wget --tries 10 -O $local_path $remote_path || failed=true
+    else
+        failed=true
```
Print an error message saying the tool couldn't be found?

natemcmaster replied:
That was added already. Line 176 below.
